Fix DNSName wildcard encoding for NameConstraints

intgr opened the pull request:
Previously '.example.com' would get normalised to 'example.com', making it impossible to add wildcard NameConstraints.

reaperhulk commented:
Jenkins, ok to test.

reaperhulk commented on tests/test_x509_ext.py:
```diff
@@ -3526,3 +3529,34 @@ def test_invalid_certificate_policies_data(self, backend):
+        permitted = [u'.example.org', u'.\xf5\xe4\xf6\xfc.example.com',
+                     u'x2603.example.net']
+        private_key = RSA_KEY_2048.private_key(backend)
+        builder = _make_certbuilder(private_key)
```
`_make_certbuilder` is likely a good idea to reduce some duplication across our tests, but since it's only in use in this one place I'd prefer that you inlined it here. I'd be happy to review a PR that uses this technique to simplify the tests in the future though!

reaperhulk commented on tests/test_x509_ext.py:
```diff
@@ -3526,3 +3529,34 @@ def test_invalid_certificate_policies_data(self, backend):
+                            excluded_subtrees=[]), True)
+
+        cert = builder.sign(private_key, hashes.SHA1(), backend)
+        result = [x.value for x in cert.extensions[0].value.permitted_subtrees]
```
Could you change this to use `get_extension_by_class(NameConstraints).value.permitted_subtrees`? While we do support indexing in this fashion I'd prefer it to be a bit more clear.

reaperhulk commented:
Sorry about the delay in reviewing this. Our decision to do unicode decoding continuously haunts me... This mostly looks good modulo the two comments I left!

intgr commented:
Updated the branch. I moved the new test into the `TestNameConstraintsExtension` class, which in hindsight seems more appropriate. And I added a new certbuilder test in `TestSubjectAlternativeName` because I also touched the code path dealing with SANs. So now there's a good reason to keep the `_make_certbuilder` function. :)

reaperhulk commented:
Thanks for your work on this! If you'd like to add your name to `AUTHORS.rst` feel free to send a PR!

intgr commented:
Oops, now I realize I had botched that commit message. Didn't see it before :(
